File: app/src/main/java/com/teixeira/subtitles/preferences/fragments/PreferencesFragment.java
package com.teixeira.subtitles.preferences.fragments;

import android.os.Bundle;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import com.teixeira.subtitles.App;
import com.teixeira.subtitles.BuildConfig;
import com.teixeira.subtitles.R;

public class PreferencesFragment extends PreferenceFragmentCompat {

  public static final String FRAGMENT_TAG = "PreferencesFragment";

  @Override
  public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
    setPreferencesFromResource(R.xml.preferences, rootKey);

    Preference generalPref = findPreference("pref_general");
    generalPref.setOnPreferenceClickListener(
        pref -> {
          changePreferenceFragment(new GeneralPreferencesFragment());
          return true;
        });

    Preference developmentPref = findPreference("pref_development");
    developmentPref.setOnPreferenceClickListener(
        pref -> {
          changePreferenceFragment(new DevelopmentPreferencesFragment());
          return true;
        });

    Preference githubPref = findPreference("pref_github");
    githubPref.setOnPreferenceClickListener(
        pref -> {
          App.getInstance().openUrl(App.APP_REPO_URL);
          return true;
        });

    Preference versionPref = findPreference("pref_version");
    versionPref.setSummary(createVersionText());
  }

  private String createVersionText() {
    String buildVersion = BuildConfig.VERSION_NAME;
    String buildType = BuildConfig.BUILD_TYPE;

    return buildVersion.concat(" (".concat(buildType.concat(")")));
  }

  private void changePreferenceFragment(PreferenceFragmentCompat fragment) {
    getParentFragmentManager()
        .beginTransaction()
        .replace(R.id.container, fragment)
        .addToBackStack(null)
        .commit();
  }
}

File: app/src/main/java/com/teixeira/subtitles/preferences/fragments/PreferencesFragment.kt
package com.teixeira.subtitles.preferences.fragments

import android.os.Bundle
import androidx.activity.OnBackPressedCallback
import androidx.fragment.app.viewModels
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.teixeira.subtitles.App
import com.teixeira.subtitles.BuildConfig
import com.teixeira.subtitles.R
import com.teixeira.subtitles.viewmodels.PreferencesViewModel

class PreferencesFragment : PreferenceFragmentCompat() {

  companion object {
    const val FRAGMENT_TAG = "PreferencesFragment"
  }

  private val preferencesViewModel by viewModels<PreferencesViewModel>()

  private val onBackPressedCallback =
    object : OnBackPressedCallback(false) {
      override fun handleOnBackPressed() {
        val currentPreferencesId = preferencesViewModel.currentPreferencesId
        if (currentPreferencesId != R.xml.preferences) {
          preferencesViewModel.currentPreferencesId = R.xml.preferences
        }
      }
    }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    requireActivity().onBackPressedDispatcher.addCallback(onBackPressedCallback)
  }

  override fun onResume() {
    super.onResume()
    onBackPressedCallback.isEnabled = preferencesViewModel.currentPreferencesId != R.xml.preferences
  }

  override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
    preferencesViewModel.observeCurrentPreferencesId(this) {
      onBackPressedCallback.isEnabled = it != R.xml.preferences
      setPreferencesFromResource(it, rootKey)
      onPreferencesIdChange(it)
    }
  }

  private fun onPreferencesIdChange(preferencesId: Int) {
    if (preferencesId == R.xml.preferences) {
      findPreference<Preference>("pref_general")?.setOnPreferenceClickListener { _ ->
        preferencesViewModel.currentPreferencesId = R.xml.preferences_general
        true
      }

      findPreference<Preference>("pref_development")?.setOnPreferenceClickListener { _ ->
        preferencesViewModel.currentPreferencesId = R.xml.preferences_development
        true
      }

      findPreference<Preference>("pref_github")?.setOnPreferenceClickListener { _ ->
        App.getInstance().openUrl(App.APP_REPO_URL)
        true
      }

      findPreference<Preference>("pref_version")?.setSummary(createVersionText())
    }
  }

  private fun createVersionText(): String {
    val buildVersion = BuildConfig.VERSION_NAME
    val buildType = BuildConfig.BUILD_TYPE
    return "$buildVersion ($buildType)"
  }
}
